core/solver/batch_solver.ipp

```diff
@@ -0,0 +1,146 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_SOLVER_BATCH_SOLVER_IPP_
+#define GKO_CORE_SOLVER_BATCH_SOLVER_IPP_
+
+
+#include <ginkgo/core/log/batch_convergence.hpp>
+#include <ginkgo/core/solver/batch_solver.hpp>
+
+
+#include "core/log/batch_logging.hpp"
+#include "core/matrix/batch_csr_kernels.hpp"
+#include "core/matrix/batch_dense_kernels.hpp"
+
+
+namespace gko {
+namespace solver {
+namespace batch {
+
+
+GKO_REGISTER_OPERATION(pre_diag_scale_system, batch_csr::pre_diag_scale_system);
+GKO_REGISTER_OPERATION(vec_scale, batch_dense::batch_scale);
+
+
+}
+
+
+struct BatchInfo {
+    //gko::log::BatchLogData<ValueType> logdata;
+    void* logdata;
+};
+
+
+template <typename ConcreteSolver, typename PolymorphicBase>
+void EnableBatchSolver<ConcreteSolver, PolymorphicBase>::apply_impl(const BatchLinOp* b,
+                                          BatchLinOp* x) const
```

Comment:
As far as I understand this, you implement the function of the Mixin (class `EnableBatchSolver`) in each cpp file corresponding to `ConcreteSolver`. That is quite hidden.
Why is it not possible to have this functionality directly inside the header? You would probably need additional functions to call kernels like `pre_diag_scale_system`, but you should be able to make them virtual and then implement them in each solver or directly inside `BatchDense` and `BatchCsr`.

Reply:
The point is to not repeat common stuff for each batch solver. I faced issues earlier with inconsistent core code between different solvers - that's not a good idea. In general, it's difficult to move this to the interface header file because it will then need non-interface headers. However, it might actually be possible after my next PR (not yet created), related to scaling and `BatchDiagonal`, that will replace some of the kernel calls with interface function calls. Let's defer this discussion till then. While it's not completely obvious that this function will have "multiple implementations", the fact that it's templated on the `ConcreteSolver` does imply that, I guess.